pkg/engine/lifecycletest/step_generator_test.go

```diff
@@ -136,3 +137,235 @@ func TestSecretMasked(t *testing.T) {
+						return "my-resource-id", news, resource.StatusOK, nil
+					},
+				}).
+				RunUpdate(func(info plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+					_, _, _, err := monitor.RegisterResource("pkgA:m:typA", "resA", true)
+					assert.NoError(t, err)
+					return nil
+				}).
+				Then(func(snap *deploy.Snapshot, err error) {
+					assert.NoError(t, err)
+
+					assert.NotNil(t, snap)
+					assert.Nil(t, snap.VerifyIntegrity())
+					assert.Len(t, snap.Resources, 2)
+					assert.Equal(t, resource.URN("urn:pulumi:test::test::pkgA:m:typA::resA"), snap.Resources[1].URN)
+					assert.True(t, !snap.Resources[1].External)
+				})
+		})
+}
+
+func TestInitErrorsStep(t *testing.T) {
+	t.Parallel()
+
+	// Create new resource for this snapshot.
+	NewTestBuilder(t, &deploy.Snapshot{
+		Resources: []*resource.State{
+			{
+				Type:   "pulumi:providers:pkgA",
+				URN:    "urn:pulumi:test::test::pulumi:providers:pkgA::default",
+				Custom: true,
+				Delete: false,
+				ID:     "935b2216-aec5-4810-96fd-5f6eae57ac88",
+			},
+			{
+				Type:     "pkgA:m:typA",
+				URN:      "urn:pulumi:test::test::pkgA:m:typA::resA",
+				Custom:   true,
+				ID:       "my-resource-id",
+				Provider: "urn:pulumi:test::test::pulumi:providers:pkgA::default::935b2216-aec5-4810-96fd-5f6eae57ac88",
+				InitErrors: []string{
+					`errors should yield an empty update to "continue" awaiting initialization.`,
+				},
+			},
+		},
+	}).
+		WithProvider("pkgA", "1.0.0", &deploytest.Provider{
+			CreateF: func(urn resource.URN, news resource.PropertyMap, timeout float64,
+				preview bool,
+			) (resource.ID, resource.PropertyMap, resource.Status, error) {
+				return "my-resource-id", news, resource.StatusOK, nil
+			},
+		}).
+		RunUpdate(func(info plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+			_, _, _, err := monitor.RegisterResource("pkgA:m:typA", "resA", true)
+			assert.NoError(t, err)
+			return nil
+		}).
+		Then(func(snap *deploy.Snapshot, err error) {
+			assert.NoError(t, err)
+
+			assert.NotNil(t, snap)
+			assert.Nil(t, snap.VerifyIntegrity())
+			assert.Len(t, snap.Resources, 2)
+			assert.Equal(t, resource.URN("urn:pulumi:test::test::pkgA:m:typA::resA"), snap.Resources[1].URN)
+			assert.Empty(t, snap.Resources[1].InitErrors)
+		})
+}
+
+func TestReadNilOutputs(t *testing.T) {
+	t.Parallel()
+
+	const resourceID = "my-resource-id"
+	NewTestBuilder(t, nil).
+		WithProvider("pkgA", "1.0.0", &deploytest.Provider{
+			ReadF: func(urn resource.URN, id resource.ID,
+				inputs, state resource.PropertyMap,
+			) (plugin.ReadResult, resource.Status, error) {
+				return plugin.ReadResult{}, resource.StatusOK, nil
+			},
+		}).
+		RunUpdate(func(info plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+			_, _, err := monitor.ReadResource("pkgA:m:typA", "resA", resourceID, "", nil, "", "", "")
+			assert.Error(t, err)
+			assert.ErrorContains(t, err, "resource 'my-resource-id' does not exist")
```

Comment:
This and below are other cases where `assert.ErrorContains` alone is sufficient -- no need to also have `assert.Error`.